lib/WordDiff.ml

```diff
@@ -0,0 +1,147 @@
+open Nottui
+module W = Nottui_widgets
+
+type word_diff =
+  | WDeleted of string array
+  | WAdded of string array
+  | WEqual of string array
+
+type line_change =
+  | CommonWord of string
+  | AddedWord of string
+  | DeletedWord of string
+  | ModifiedDiff of word_diff list
+
+type hunk = {
+  mine_start : int;
+  mine_len : int;
+  their_start : int;
+  their_len : int;
+  lines : line_change list;
+}
+
+let string_to_words s = Array.of_list (String.split_on_char ' ' s)
+let words_to_string words = String.concat " " (Array.to_list words)
+let longest xs ys = if List.length xs > List.length ys then xs else ys
+
+let lcs xs' ys' =
+  let xs = Array.of_list xs' and ys = Array.of_list ys' in
+  let n = Array.length xs and m = Array.length ys in
+  let a = Array.make_matrix (n + 1) (m + 1) [] in
+  for i = n - 1 downto 0 do
+    for j = m - 1 downto 0 do
+      a.(i).(j) <-
+        (if xs.(i) = ys.(j) then xs.(i) :: a.(i + 1).(j + 1)
+         else longest a.(i).(j + 1) a.(i + 1).(j))
+    done
+  done;
+  a.(0).(0)
+
+let diff_words s1 s2 =
+  let words1 = Array.to_list (string_to_words s1) in
+  let words2 = Array.to_list (string_to_words s2) in
+  let common = lcs words1 words2 in
+  let rec construct_diff w1 w2 lcs acc =
+    match (w1, w2, lcs) with
+    | [], [], [] -> List.rev acc
+    | x :: xs, y :: ys, z :: zs when x = z && y = z ->
+        construct_diff xs ys zs (WEqual [| z |] :: acc)
+    | x :: xs, ys, lcs when not (List.mem x lcs) ->
+        construct_diff xs ys lcs (WDeleted [| x |] :: acc)
```

Comment:
Why do you need to check whether $x\in\texttt{lcs}$? This seems very costly, and wrong!